fedimint-bin-tests/src/federation.rs

```diff
@@ -0,0 +1,320 @@
+use std::collections::HashMap;
+
+use anyhow::Context;
+use fedimint_core::core::LEGACY_HARDCODED_INSTANCE_ID_WALLET;
+use fedimint_core::db::mem_impl::MemDatabase;
+use fedimint_core::task::TaskGroup;
+use fedimint_wallet_client::config::WalletClientConfig;
+use tokio::fs;
+use tokio::sync::mpsc::{self, Receiver, Sender};
+
+use super::*; // TODO: remove this
+
+/// Create a map of environment variables which fedimintd and DKG can use,
+/// but which can't be defined by `build.sh` because multiple of these daemons
+/// run concurrently with different values.
+///
+/// We allow ranges of 10 ports for each fedimintd / dkg instance starting from
+/// 18173. Each port needed is incremented by 1 within this range.
+///
+/// * `id` - ID of the server. Used to calculate port numbers.
+fn fedimint_env(id: usize) -> anyhow::Result<HashMap<String, String>> {
+    let base_port = 8173 + 10000;
+    let p2p_port = base_port + (id * 10);
+    let api_port = base_port + (id * 10) + 1;
+    let ui_port = base_port + (id * 10) + 2;
+    let cfg_dir = env::var("FM_CFG_DIR")?;
+    Ok(HashMap::from_iter([
+        ("FM_BIND_P2P".into(), format!("127.0.0.1:{p2p_port}")),
+        (
+            "FM_P2P_URL".into(),
+            format!("fedimint://127.0.0.1:{p2p_port}"),
+        ),
+        ("FM_BIND_API".into(), format!("127.0.0.1:{api_port}")),
+        ("FM_API_URL".into(), format!("ws://127.0.0.1:{api_port}")),
+        ("FM_LISTEN_UI".into(), format!("127.0.0.1:{ui_port}")),
+        (
+            "FM_FEDIMINT_DATA_DIR".into(),
+            format!("{cfg_dir}/server-{id}"),
+        ),
+        ("FM_PASSWORD".into(), format!("pass{id}")),
+    ]))
+}
+
+pub struct Federation {
+    // client is only for internal use, use cli commands instead
+    client: Arc<UserClient>,
+    pub members: Vec<Fedimintd>,
```

Comment:
For a one-off access like that a `pub fn members(&self) -> &[Fedimintd` is often better.